// @algo-builder/web/build/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.runtimeGenesisHash = exports.betanetGenesisHash = exports.testnetGenesisHash = exports.mainnetGenesisHash = exports.betanetURL = exports.testnetURL = exports.mainnetURL = exports.ALGORAND_SIGN_TRANSACTION_REQUEST = exports.WAIT_ROUNDS = exports.reDigit = exports.MAX_UINT64 = exports.MIN_UINT64 = void 0;
exports.MIN_UINT64 = 0n;
exports.MAX_UINT64 = 0xffffffffffffffffn;
exports.reDigit = /^\d+$/;
exports.WAIT_ROUNDS = 10; // for transaction confirmation
// wallect connect
exports.ALGORAND_SIGN_TRANSACTION_REQUEST = "algo_signTxn";
exports.mainnetURL = "https://node.algoexplorerapi.io";
exports.testnetURL = "https://node.testnet.algoexplorerapi.io";
exports.betanetURL = "https://node.betanet.algoexplorerapi.io";
//official networks https://developer.algorand.org/docs/get-details/algorand-networks/
exports.mainnetGenesisHash = "wGHE2Pwdvd7S12BL5FaOP20EGYesN73ktiC1qzkkit8=";
exports.testnetGenesisHash = "SGO1GKSzyE7IEPItTxCByw9x8FmnrCDexi9/cOUJOiI=";
exports.betanetGenesisHash = "mFgazF+2uRS1tMiL9dsj01hJGySEmPN28B/TjjvpVW0=";
//runtime network
exports.runtimeGenesisHash = "runtime";
